File: LC54.py
class Solution:
    def spiralOrder(self, mat: List[List[int]]) -> List[int]:
        l = 0
        r = len(mat[0]) - 1
        t = 0
        b = len(mat) - 1
        res = []

        while l < r and t < b:
            for i in range(l,r):
                res.append(mat[t][i])

            for i in range(t,b):
                res.append(mat[i][r])
            
            for i in range(r,l,-1):
                res.append(mat[b][i])
            
            for i in range(b,t,-1):
                res.append(mat[i][l])
            
            l += 1; r -= 1
            t += 1; b -= 1
        
        if l == r:
            while t <= b:
                res.append(mat[t][r])
                t += 1
        
        if t == b:
            while l <= r:
                res.append(mat[b][l])
                l += 1
        
        return res

File: LC54.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        vector<int> res;

        int l = 0, b = matrix.size()-1, t = 0, r = matrix[0].size()-1;
        while(l < r && t < b){
            for(int i = l; i < r; i++)
                res.push_back(matrix[t][i]);
            for(int i = t; i < b; i++)
                res.push_back(matrix[i][r]);
            for(int i = r; i > l; i--)
                res.push_back(matrix[b][i]);
            for(int i = b; i > t; i--)
                res.push_back(matrix[i][l]);
            
            l++; r--;
            t++; b--;
        }

        if(t == b){
            while(l <= r)
                res.push_back(matrix[t][l++]);
        }

        if(l == r){
            while(t <= b)
                res.push_back(matrix[t++][r]);
        }

        return res; 
    }
};
